### DISCUSSION TITLE
ILI9341 + NodeMcu32 -> 8Bit-Parallel (works) + Touch (doesn't work correct)

### DISCUSSION THREAD

**Discussion by nitrolack** [Bodmer/TFT_eSPI#2389] [Q&A - Touch] [answered]
I got the display working and it displays the color test correctly, but when i want to use the adafruit_touchscreen library it outputs negative pressures and negative values overall. Tried many stuff around, even tried to debug it and so, but i do know, that it have something to do with my settings.
The ILI9341 and the touchscreen (probably xpt2046) share pins together.

The display: https://www.reichelt.de/arduino-shield-display-2-4-touch-320-x-240-pixel-ili9341-ard-shd-2-4td-p282495.html
The esp: https://www.reichelt.de/nodemcu-esp32-wifi-und-bluetooth-modul-debo-jt-esp32-p219897.html?&trstct=pos_1&nbc=1

My wiring:
#define TFT_PARALLEL_8_BIT
#define ILI9341_DRIVER
//  RS/DC and D0 & CS and D1
// ESP32 pins used for the parallel interface TFT
#define TFT_CS   33  // Chip select control pin
#define TFT_DC   15  // Data Command control pin - must use a pin in the range 0-31
#define TFT_RST  32  // Reset pin
#define TFT_WR    4  // Write strobe control pin - must use a pin in the range 0-31
#define TFT_RD    2
#define TFT_D0   14  // Must use pins in the range 0-31 for the data bus
#define TFT_D1   12  // so a single register write sets/clears all bits
#define TFT_D2   26
#define TFT_D3   25
#define TFT_D4   17
#define TFT_D5   16                                                                                                                                                                                                   
#define TFT_D6   22       
#define TFT_D7   21 

I already debugged a bit and took the getPoint function apart and i found some Error:
int NUMSAMPLES = 3;
    int x1, y1, z1;
    int samples[NUMSAMPLES];
    uint8_t i, valid;
    valid = 1;
    pinMode(YP, INPUT);
    pinMode(YM, INPUT);
    digitalWrite(YP, LOW);
    digitalWrite(YM, LOW);
    pinMode(XP, OUTPUT);
    pinMode(XM, OUTPUT);
    digitalWrite(XP, HIGH);
    digitalWrite(XM, LOW);
    for (i = 0; i < NUMSAMPLES; i++) {
        samples[i] = analogRead(YP);
        //delay(10);
    };
    insert_sort(samples, NUMSAMPLES);
    Serial.println(samples[0]);
    Serial.println(samples[1]);
    Serial.println(samples[2]);
    Serial.println(samples[3]);
    x1 = (1023 - samples[NUMSAMPLES / 2]);

When i print the samples array, when i touch nothing,  it outputs for samples[0] around 3000 and for 1&2 4095.
When i touch the display they can reach from 500 in the one corner to 3300 in the other corner.
So on it has something to do with the analogRead and that these samples aren't standardized in the area from 500 till 3300.
Does anyone has an idea, how i can fix this and the library of adafruit to get it working?

**Comment by Bodmer** [comment, marked as answer]
The touch screen interface is anlaogue (no XPT2046) and the touch pins share with the digital pins to the TFT controller, so this complicates matters.

There is an adapted Adafruit library that handles this, note that this assumes the ESP32 is wired with additional connections so some TFT shield pins link to multiple ESP32 pins.

See the end of the ReadMe section [here](https://github.com/Bodmer/TFT_eSPI#8-bit-parallel-support) for a link to the adapted library and the connections made. You will need to figure out how the UNO board modifications translate to your setup.

**Reply by nitrolack** [reply]
Thank you for your fast reply bodmer.
Gonna try this out. Now i understand it a bit more.
So it's a classical resistive touch.
I gonna go and try to get it working and post my accomplishments in this discussion.

**Reply by nitrolack** [reply]
UPDATE:
Got it working, but it's always mirrored, nevertheless which orientation i choose.

**Reply by Bodmer** [reply]
Did you run the calibration sketch linked to in the touch library's ReadMe?

**Reply by nitrolack** [reply]
I just switched some coords in the definition and now it's not mirrored anymore.
Now i just have to calibrate it, either by hand or with the script, but i dont wan't to use the mcufriend script, cause i use tft_espi.

**Reply by nitrolack** [reply]
Got it all working with just trying values per hand with touching the display, but i got another problem.
I gonna start a new discussion for it.
